Storage:  Fixes GPT alternative header not being at the end of the VM block device

tomponline opened the pull request:
Addresses issue https://github.com/lxc/lxd/issues/6929

There are 2 issues here:

1. We need to move the GPT alternative header to the end of the disk after unpacking a VM image to a disk file. This is what the GPT standard requires so it knows where to find the backup header. This change needs to be added to each VM capable driver:

- [x] ZFS
- [x] LVM
- [x] DIR
- [x] BTRFS
- [x] CEPH

2. When cloning a ZFS volume from a snapshot, because snapshots don't have the `volmode` property, the property is instead inherited from parent (which LXD currently doesn't modify from default), and so when the new volume is created ZFS then probes the partitions on the disk and creates a device for each partition. This probing causes the host kernel to detect the the problem GPT header and logs warnings to the kernel log.

There is also potentially a bug in ZFS itself, because specifying `volmode=none` on the `zfs clone` command is not respected and instead it still uses the parent setting, even though `zfs get volmode` will show the correct setting on the cloned volume.

Although fixing the first issue will fix the second issue, it is an indication of a wider problem in that we do not want the partitions of the VM image to be presented on the host.

tomponline commented:
@stgraber just putting this up now to get jenkins to test what I've done so far, and to check my top-level approach with you.

stgraber commented on lxd/storage/drivers/utils.go:
```diff
@@ -573,3 +575,26 @@ func copyDevice(inputPath, outputPath string) error {
+// moveGPTAltHeader moves the GPT alternative header to the end of the disk device supplied.
+// If the device supplied is not detected as not being a GPT disk then no action is taken and nil is returned.
+func moveGPTAltHeader(devPath string) error {
+	_, err := shared.RunCommand("sgdisk", "--move-second-header", devPath)
```
How about only doing this if sgdisk can be found and logging a warning if not?

Not sure how common that tool is.

tomponline replied:
Yep will do.

stgraber commented on lxd/storage/drivers/driver_zfs_volumes.go:
```diff
@@ -130,6 +130,14 @@ func (d *zfs) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Oper
+					d.logger.Debug("Moved GPT alternative header to end of disk")
+				} else if err != nil && err != errNonGPTDisk {
+					return err
+				}
```
So I'm not completely sure about that part, do we really want to do it in every driver whenever we alter the storage size (also need to deal with growing the volume) or can we maybe do this from the backend in one central place (all VM instance creations & when updating the size of a VM).

stgraber commented:
@tomponline we also still need the zfs workaround for volmode=none, right?

I may quickly send you that one as I've done a fair number of storage patches already :)

tomponline commented:
Yes I still need to do the parent volmode patch and change for new pools.

As for doing the gpt move in the backendlxd, I considered that, but ruled it out as it would require calling volume Mount so we can be sure the device actually exists for zfs, which seemed inefficient.  Is there a reason we use volmode none rather then volmode dev so the devices exist all the time like the other storage drivers?

stgraber commented:
https://github.com/lxc/lxd/pull/6933

stgraber commented:
@tomponline to mimic our dataset behavior where we don't keep anything mounted/visible unless needed. ZFS doesn't always scale amazingly well when everything is active.

Okay, so if we do it at the driver layer, we need to make sure we do it when:
 - filling a new volume
 - copying from an existing volume with a size change
 - applying a new size to the volume

tomponline commented:
Yep sounds good.

tomponline commented:
> * copying from an existing volume with a size change

That middle option is interesting, looking at the code for (at least) ZFS and LVM, neither drivers support that option (i.e creating an instance from an image using a profile with a root disk `size` option specified), the existing image volume is just used as is.

tomponline commented:
@stgraber I've taken a closer look at LVM, and that does resize the snapshot taken when using optimised volumes, but only by specifying the pool's `volume.size` settings, its not taken from the instance's root disk profile `size` property (not sure if thats valid), nor can I see a way to specify it at create time. Although it should work OK.

Will double check ZFS works the same way as LVM (by using SetVolumeQuota), and if so then GPT header moving will also work fine.

tomponline commented:
@stgraber no, so ZFS driver doesn't currently take into account changes in the existing snapshotted image volume and the `volume.size` property on the pool. But that is a separate bug will fix in separate PR.

tomponline commented:
@stgraber also noticed that the `dir` driver doesn't support VM resize at all currently. Only initial sizing based on `volume.size`.

tomponline commented:
@stgraber I've added GPT moving to the places I can for now, and I've identified a few VM related size issues in all drivers except LVM, have logged those as individual task points in our VM work trello card.

stgraber commented:
Ok, would probably be worth fixing those ahead of the project storage bits as I'd like resizing to behave in 3.22.

tomponline commented:
@stgraber I've fixed the `dir` and `btrfs` driver's support for `SetVolumeQuota` for VM image files. Before now, the `btrfs` driver was still applying BTRFS quota when using block type volumes, however I've now changed it so that block volumes only resize the block file and do not apply BTRFS quota at all. Is this what you would expect (i.e the same behaviour as `dir` driver)?

Additionally I have some questions about the `btrfs` driver:

1. From what I can see, the `CreateVolume` and `CreateVolumeFromCopy` functions do not apply the volume's `size` property as a quota for filesystem volumes (unlike `dir` driver does). Is this intentional or an oversight?

2. Related to Q1, in order to correctly resize the VM block file during `CreateVolume` and `CreateVolumeFromCopy` I would like to call `SetVolumeQuota` during those functions, however that will have the effect of applying a quota for filesystem volumes which doesn't appear to happen right now.

stgraber commented:
Yeah, that's good for btrfs.

stgraber commented:
Lack of quota is an oversight.

tomponline commented:
ZFS quota applied when creating from copy.

tomponline commented:
Add BTRFS quota fixes.
